### PR TITLE
Refactor tx filters

### LINKED ISSUE #3623: Unify ITxFilter and IIncomingTxFilter
Provide better code and messages to RPC requests as the ones from ITxFilter's are now just 'Filtered'.

### PR DESCRIPTION
**Pull request by marcindsobczak**
Resolves #3623 

## Changes:
Changing ITxFilter to be like IIncomingTxFilter (returns AddTxResult? instead of string).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by kjazgar — APPROVED** [review]

File: src/Nethermind/Nethermind.Consensus/Transactions/NullTxFilter.cs [hunk 1]
```diff
@@ -16,12 +16,13 @@
...
 
 using Nethermind.Core;
+using Nethermind.TxPool;
 
 namespace Nethermind.Consensus.Transactions
 {
     public class NullTxFilter : ITxFilter
     {
-        public (bool Allowed, string Reason) IsAllowed(Transaction tx, BlockHeader parentHeader) => (true, string.Empty);
+        public (bool Allowed, AddTxResult? Reason) IsAllowed(Transaction tx, BlockHeader parentHeader) => (true, null);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Transactions/NullTxFilter.cs:25]: As we have AddTxResult.Added, we don't need the tuple now, we can simplify this a lot.

File: src/Nethermind/Nethermind.TxPool/AddTxResult.cs [hunk 2]
```diff
@@ -71,6 +76,11 @@ public enum AddTxResult
         /// </summary>
         Invalid,
         
+        /// <summary>
+        /// Permission denied for this tx type.
+        /// </summary>
+        PermissionDeniedForThisTxType,
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/AddTxResult.cs:82]: just PermissionDenied

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/AddTxResult.cs:82]: Also I don't love the fact that POSDAO based specific result now got into here. Maybe we could make AddTxResult a struct tha would have:
1) Number
2) Code
3) Optional custom message?

Then other modules when extending TxPool could return their won results.

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]
We still have both ITxFilter and IIncomingTxFilter, we should unify and simplify. Then delete the obsolete one.

This is good first step but not enough for proper refactor.

File: src/Nethermind/Nethermind.Blockchain/TxFilterAdapter.cs [hunk 3]
```diff
@@ -37,23 +37,23 @@ public TxFilterAdapter(IBlockTree blockTree, ITxFilter txFilter, ILogManager log
...
             if (tx is not GeneratedTransaction)
             {
                 BlockHeader parentHeader = _blockTree.Head?.Header;
-                if (parentHeader == null) return (true, null);
+                if (parentHeader == null) return AcceptTxResult.Accepted;
 
-                (bool accepted, string? reason) = _txFilter.IsAllowed(tx, parentHeader);
-                if (reason is not null)
+                AcceptTxResult acceptTxResult = _txFilter.IsAllowed(tx, parentHeader);
+                if (!acceptTxResult.Equals(AcceptTxResult.Accepted))
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/TxFilterAdapter.cs:48]: implicit bool cast

File: src/Nethermind/Nethermind.Baseline.Test/BaselineTreeTracker_ReorganizationTests.cs [hunk 4]
```diff
@@ -90,8 +90,8 @@ public async Task Tree_tracker_reorganization([ValueSource(nameof(Reorganization
                 transaction.SenderAddress = key.Address;
                 ecdsa.Sign(key, transaction, true);
                 transaction.Hash = transaction.CalculateHash();
-                AddTxResult result = testRpc.TxPool.SubmitTx(transaction, TxHandlingOptions.None);
-                if (result != AddTxResult.Added)
+                AcceptTxResult result = testRpc.TxPool.SubmitTx(transaction, TxHandlingOptions.None);
+                if (!result.Equals(AcceptTxResult.Accepted))
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Baseline.Test/BaselineTreeTracker_ReorganizationTests.cs:94]: implicit bool cast

File: src/Nethermind/Nethermind.TxPool/AcceptTxResult.cs [hunk 5]
```diff
@@ -13,83 +13,100 @@
...
+            Code = code;
+            Message = message;
+        }
+
+        public static implicit operator bool(AcceptTxResult result) => result.Id == 0;
+        public AcceptTxResult WithMessage(string message) => new(Id, Code, message);
+        public static bool operator ==(AcceptTxResult a, AcceptTxResult b) => a.Equals(b);
+        public static bool operator !=(AcceptTxResult a, AcceptTxResult b) => !(a == b);
+        public override bool Equals(object? obj) => Equals((AcceptTxResult)obj);
+        public bool Equals(AcceptTxResult result) => Id == result.Id;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/AcceptTxResult.cs:108]: add implementation IEquatable<AcceptTxResult>

File: src/Nethermind/Nethermind.TxPool/AcceptTxResult.cs [hunk 6]
```diff
@@ -13,83 +13,100 @@
...
+        private string? Message { get; }
+
+        public AcceptTxResult(int id, string code, string? message = null)
+        {
+            Id = id;
+            Code = code;
+            Message = message;
+        }
+
+        public static implicit operator bool(AcceptTxResult result) => result.Id == 0;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/AcceptTxResult.cs:103]: can we go with result.Id == Accepted.Id ?

**Review by LukaszRozmej — APPROVED** [review]
We could rename "result" to "accepted" or "allowed" in many places now so it would be more readable with those if's.

File: src/Nethermind/Nethermind.TxPool/AcceptTxResult.cs (hunk 6 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/AcceptTxResult.cs:103]: Or maybe make it a const here and there

**Comment by prashantkumar1982** [comment]
This PR broke Chainlink integration.
The refactoring of error message broke our regular expression matching logic, which tells us why a Tx failed, and makes our StateMachine work.

A partial fix that Chainlink needed on our side is here: https://github.com/smartcontractkit/chainlink/pull/7989

@marcindsobczak in future, please keep this in mind, and possibly, could you notify Chainlink before changing error response strings?

**Comment by marcindsobczak** [comment]
Thank you @prashantkumar1982 for a message. We highly appreciate such a feedback, although we don't have a knowledge how your setup works, so it would be hard to catch which changes can possibly affect it. Possible solution could be to add integration tests which we can run on our CI. Could you please provide us such a tests? We would love to add it to our codebase and then it will be easy to remember to ping you every time when we change something and tests start failing

**Comment by dimriou** [comment]
Hey @marcindsobczak , thanks for the suggestions. We'll discuss this internally and let you know. 

For reference, here (https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/client/errors.go#L170) you can find the regexes we use to match Nethermind's RPC responses. Depending on each response, we make a different decision on how to handle our pending transactions. As @prashantkumar1982 mentioned, if the regexes change this means that our Nodes will not be able to handle some of the responses, thus affecting our pending transactions.

**Pull request merged**